gateway/gateway.go

```diff
@@ -118,6 +125,10 @@ func New(listenAddr string, upstreamer Upstreamer, options ...Option) (Gateway,
 		gatewayConfig: cfg,
 	}
 
+	if _, ok := s.upstreamer.(ResponseTimeCollector); ok {
+		s.implementResponseTimeCollector = true
```

Comment:
Store the upstreamer directly as a LatencyBasedUpstreamer to avoid additional casts